service/matching/matching_engine_test.go

```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tl,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: "unknown",
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil)
+
+	_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:            namespaceId.String(),
+		Execution:              &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+		ScheduledEventId:       0,
+		TaskQueue:              &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+		ForwardedSource:        "somewhere", // force sync match only
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: "unknown",
+			},
+		},
+	})
+	s.ErrorIs(err, errRemoteSyncMatchFailed)
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_tasks+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
+	s.Equal(int64(1), unknownCtr.Value())
+}
+
+func (s *matchingEngineSuite) TestPersistUnknownBuildId_Match() {
+	namespaceId := namespace.ID(uuid.New())
+	tl := "makeToast"
+
+	scope := tally.NewTestScope("test", nil)
+	s.matchingEngine.metricsHandler = metrics.NewTallyMetricsHandler(metrics.ClientConfig{}, scope)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tl,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: "unknown",
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil).AnyTimes() // might get called again on dispatch from spooled
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+			NamespaceId:            namespaceId.String(),
+			Execution:              &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+			ScheduledEventId:       123,
+			TaskQueue:              &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+			ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+			// do not set ForwardedSource, allow to go to db
+			VersionDirective: &taskqueue.TaskVersionDirective{
+				Value: &taskqueue.TaskVersionDirective_BuildId{
+					BuildId: "unknown",
+				},
+			},
+		})
+		s.NoError(err)
+		wg.Done()
+	}()
+
+	go func() {
+		time.Sleep(10 * time.Millisecond)
```

Comment:
Could you comment why we need a sleep here? Also, could you run with a high `-count` to verify that this isn't flaky? Not that we have to eradicate time.Sleep from all tests; it might be prohibitively difficult to avoid in this case, but I want to at least make sure it isn't flaky.